src/librustc/traits/error_reporting.rs

```diff
@@ -940,7 +940,13 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 bug!("overflow should be handled before the `report_selection_error` path");
             }
         };
-        self.note_obligation_cause(&mut err, obligation);
```

Comment:
what about the other call sites to `note_obligation_cause`? Wouldn't they also benefit from this new diagnostic?

Reply:
I've updated `note_obligation_cause` so that it isn't generic over the predicate and takes the `PredicateObligation` type that `note_obligation_cause_for_async_await` takes, then I can call `note_obligation_cause_for_async_await` in that function. 

There was only one call site of `note_obligation_cause` that wasn't a `PredicateObligation`, so I just called `note_obligation_cause_code` directly from there (which is what would have happened previously).